ethcore/src/engines/validator_set/contract.rs

```
@@ -14,84 +14,74 @@
-		self.validators.read().contains(address)
+		self.validators.contains(address)
 	}
 
 	fn get(&self, nonce: usize) -> Address {
-		self.validators.read().get(nonce).clone()
+		self.validators.get(nonce)
 	}
 
 	fn count(&self) -> usize {
-		self.validators.read().count()
+		self.validators.count()
 	}
 
-	fn register_call_contract(&self, client: Weak<Client>) {
-		if let Some(c) = client.upgrade() {
-			c.add_notify(self.clone());
+	fn report_malicious(&self, address: &Address) {
+		if let Some(ref provider) = *self.provider.read() {
+			match provider.report_malicious(address) {
+				Ok(_) => warn!(target: "engine", "Reported malicious validator {}", address),
+				Err(s) => warn!(target: "engine", "Validator {} could not be reported {}", address, s),
+			}
+		} else {
+			warn!(target: "engine", "Malicious behaviour could not be reported: no provider contract.")
 		}
-		{
-			*self.provider.write() = Some(provider::Contract::new(self.address, move |a, d| client.upgrade().ok_or("No client!".into()).and_then(|c| c.call_contract(a, d))));
+	}
+
+	fn report_benign(&self, address: &Address) {
+		if let Some(ref provider) = *self.provider.read() {
+			match provider.report_benign(address) {
+				Ok(_) => warn!(target: "engine", "Reported benign validator misbehaviour {}", address),
+				Err(s) => warn!(target: "engine", "Validator {} could not be reported {}", address, s),
+			}
+		} else {
+			warn!(target: "engine", "Benign misbehaviour could not be reported: no provider contract.")
 		}
-		self.update();
+	}
+
+	fn register_contract(&self, client: Weak<Client>) {
+		self.validators.register_contract(client.clone());
+		let transact = move |a, d| client
+			.upgrade()
+			.ok_or("No client!".into())
+			.and_then(|c| c.transact_contract(a, d).map_err(|e| format!("Transaction import error: {}", e)))
+			.map(|_| Default::default());
+		*self.provider.write() = Some(provider::Contract::new(self.validators.address, transact));
```

Is this `Default::default()` contract really needs to be set as `provider` when there is no `Client`? Is this a part of some real logic? Maybe it should be left as `None` and return error or fail fast with other method?